### PR TITLE
Fixed #25187 -- Made request available in authentication backends.

### PR DESCRIPTION
**Pull request by AleksejManaev**
https://code.djangoproject.com/ticket/25187

### CODE REVIEW

File: tests/auth_tests/test_auth_backends.py [hunk 1]
```diff
@@ -606,6 +606,27 @@ def test_type_error_raised(self):
...
+class AcceptsRequestBackend(object):
+    """
+    Authenticate method accepts a request parameter.
+    """
+
+    def authenticate(self, request, username=None, password=None):
+        pass
+
+
+class AcceptsRequestBackendTest(TestCase):
```
Comment (charettes) [inline tests/auth_tests/test_auth_backends.py]: Please use `SimpleTestCase` here as no database query is performed.

**Comment by charettes** [comment]
Buildbot, add to whitelist.

**Comment by timgraham** [comment]
See https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature for details about all the deprecation requirements, including documentation.

**Comment by AleksejManaev** [comment]
Now the use of the deprecation in Django itself must be removed. The ModelBackend and the RemoteUserBackend accept no request parameter. Is it ok to change the authenticate signatures to
`authenticate(self, request, username=None, password=None, **kwargs):` for ModelBackend
and `authenticate(self, request, remote_user):` for RemoteUserBackend?

**Comment by timgraham** [comment]
I think that's okay.

File: docs/internals/deprecation.txt [hunk 2]
```diff
@@ -200,6 +200,11 @@ details on these changes.
 * Support for old-style middleware using ``settings.MIDDLEWARE_CLASSES`` will
   be removed.
 
+See the :ref:`Django 1.11 release notes <deprecated-features-1.11>` for more
```
Comment (timgraham) [inline docs/internals/deprecation.txt]: Chop this -- looks like a bad merge conflict resolution.

File: docs/internals/deprecation.txt [hunk 3]
```diff
@@ -200,6 +200,11 @@ details on these changes.
...
+details on these changes.
+
+* The ``authenticate()`` method of authentication backends will require a ``request`` argument.
```
Comment (timgraham) [inline docs/internals/deprecation.txt]: Please wrap documentation lines at 79 characters.

File: django/contrib/auth/backends.py [hunk 4]
```diff
@@ -8,8 +8,11 @@ class ModelBackend(object):
     """
     Authenticates against settings.AUTH_USER_MODEL.
     """
+    def __init__(self, request=None):
```
Comment (timgraham) [inline django/contrib/auth/backends.py]: It's not clear to me why these lines are needed or why `self.request = request` is assigned in `authenticate()`. At least some tests should be added for that if they are needed.

Comment (AleksejManaev) [inline django/contrib/auth/backends.py]: The ticket says to make request available in authentication backends. I thought it also applies to the ModelBackend and RemoteUserBackend. That is why `self.request = request` is assigned in `authenticate()`.

File: django/contrib/auth/__init__.py [hunk 5]
```diff
@@ -59,22 +61,33 @@ def _get_user_session_key(request):
     return get_user_model()._meta.pk.to_python(request.session[SESSION_KEY])
 
 
-def authenticate(**credentials):
-    """
```
Comment (timgraham) [inline django/contrib/auth/__init__.py]: why remove the docstring?

File: docs/releases/1.10.txt [hunk 6]
```diff
@@ -1166,6 +1166,9 @@ Miscellaneous
   deprecated. :ref:`Adapt old, custom middleware <upgrading-middleware>` and
   use the new :setting:`MIDDLEWARE` setting.
 
+* Calls to the ``authenticate()`` method of authentication backends without a
```
Comment (timgraham) [inline docs/releases/1.10.txt]: Please move to 1.11 release notes.

File: django/contrib/auth/backends.py [hunk 7]
```diff
@@ -8,8 +8,8 @@ class ModelBackend(object):
     """
     Authenticates against settings.AUTH_USER_MODEL.
     """
-
-    def authenticate(self, username=None, password=None, **kwargs):
+    def authenticate(self, request, username=None, password=None, **kwargs):
+        self.request = request
```
Comment (timgraham) [inline django/contrib/auth/backends.py]: I'm not sure if this method should set an attribute as opposed to `authenticate()` passing `request` to the methods it calls as needed. @carljm, any thought about this?

**Comment by timgraham** [comment]
There's a "Writing an authentication backend" section in topics/auth/customizing.txt to update. Also, tests for the deprecation as described in https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature are needed.

File: docs/internals/deprecation.txt [hunk 8]
```diff
@@ -200,6 +200,9 @@ details on these changes.
 * Support for old-style middleware using ``settings.MIDDLEWARE_CLASSES`` will
   be removed.
 
+* The ``authenticate()`` method of authentication backends will require a
```
Comment (timgraham) [inline docs/internals/deprecation.txt]: Move this to the correct section.

File: django/contrib/auth/__init__.py [hunk 9]
```diff
@@ -1,11 +1,13 @@
 import inspect
 import re
+import warnings
 
 from django.apps import apps as django_apps
 from django.conf import settings
 from django.core.exceptions import ImproperlyConfigured, PermissionDenied
 from django.middleware.csrf import rotate_token
 from django.utils.crypto import constant_time_compare
+from django.utils.deprecation import RemovedInDjango20Warning
```
Comment (timgraham) [inline django/contrib/auth/__init__.py]: Django21 now

**Comment by timgraham** [comment]
Please rebase your branch and squash the commits so that the tests can run.

**Comment by timgraham** [comment]
merged in 4b9330ccc04575f9e5126529ec355a450d12e77c, thanks!

**Pull request closed without merging**